[23.05] ramips: migrate scripts for Edgerouter X to new layout

darkxst opened the pull request:
UPDATE: I have reworked this PR to instead include a migration script that can be run to upgrade from 23.05.x -> 24.x, since the below method was not suitable for inclusion in a stable release.


This PR backports required parts from PR #15194, to allow for direct sysupgrade to builds from main or future 24.x where the kernels are greater than 3MB.

The OEM layout on the Edgerouter X has two kernel slots that are 3MB each. Starting with Linux 6.1 the kernel images no longer fit into this layout.

This PR will prepare the Edgerouter X towards using a single kernel slot that is 6MB, this is achieved with the following changes:

- Adjust dts to make kernel1 6MB and drop kernel2 slot
- Patch ubnt.sh to always flash to kernel1 slot (Likewise set the flag in factory partition to make sure kernel1 slot is active)
- Set compat version to 2.0, to ensure the system has the updated ubnt.sh providing support for the new layout.

This backport pre-seeds the dts layout  and `ubnt.sh`  changes, so that on a subsequent sysupgrade the system will be fully converted to single slot 6MB kernel layout. This will allow to directly sysupgrade to builds with larger kernels with no further manual intervention for user.

- The first sysupgrade to this version will switch dts layout to 6MB kernel slot and preintstall `ubnt.sh` change. (kernel could land in either kernel1 or kernel2 slot)
- On the subsequent sysupgrade, kernel will be installed installed to kernel1 slot and system will have been converted to 6MB kernel slot.

Djfe commented:
I fail to see why this requires a major compat increase.
Can't we set the image so this creates a minor compat increase but sets the uci to a major compat increase that is compatible with the upcoming image?
I mean you want to allow sysupgrades to this patch and you want to allow sysupgrades from this patch to the upcoming release, right?

darkxst commented:
Yeh, I specifically didn't bump the image compat version here, since so long as the kernel is less than 3MB (which for 23.05 it is) its not really a breaking change.

ynezz reviewed (changes requested):
Sorry, but this kind of changes are not acceptable for a stable release.

Would it be possible to provide some kind of a migration script instead? I might be missing something why following wouldn't work?

1. wget `openwrt-24.0x.y-ramips-mt7621-ubnt_edgerouter-x-squashfs-sysupgrade.bin`
2. image signature check
3. force kernel index 0
4. untar and flash 0x300000 of `sysupgrade-ubnt_edgerouter-x/kernel` into `kernel1` partition, rest into `kernel2` partition
5. flash `sysupgrade-ubnt_edgerouter-x/root` as usual

darkxst commented:
Yes it would be possible to make a migration script like that.

Do you mean as a standalone external script or somehow integrated into sysupgrade (which I dont see how its possible, given sysupgrade scripts already exist on existing system?)

Assuming the first, steps 1-4 are very straight forward, but flashing  `sysupgrade-ubnt_edgerouter-x/root` seems somewhat complicated, atleast in how sysupgrade handles it.

ynezz commented:
> Do you mean as a standalone external script

Since this is targeting a stable release, IMO the only acceptable way would be external upgrade tool/script added under `target/linux/ramips/base-files`

> or somehow integrated into sysupgrade (which I dont see how its possible, given sysupgrade scripts already exist on existing system?)

We shouldn't touch that parts in the stable release.

darkxst commented:
> Since this is targeting a stable release, IMO the only acceptable way would be external upgrade tool/script added under `target/linux/ramips/base-files`

OK, I will have a look and see what I can come up.

darkxst committed:
ramips: mt7621: Provide script to migrate Edgerouter-X to Linux 6.6

Linux 6.6 requires a new layout to allow for kernels > 3MB.

This script will migrate existing 23.05 systems to the new layout
that will be used in snapshots and future 24.x releases:
- Download sysupgrade image
- Pivot to ramfs
- Flash kernel
- Force kernel boot flag to slot 1
- Flash rootfs (user settings are not retained)

Signed-off-by: Tim Lunn <tim@feathertop.org>

darkxst commented:
> Would it be possible to provide some kind of a migration script instead? I might be missing something why following wouldn't work?

I created a migration script, it wasn't exactly simple though. Need to pivot into ramfs to be able to flash rootfs etc. As such ended up hooking into parts of sysupgrade.

First script downloads sysupgrade file (well apart from that there are no images available yet), then pivots into ramfs using ubus sysupgrade call.

Second script runs in the ramfs and flashes kernel and rootfs.

Hope this will be acceptable for backport to 23.05?

msandber commented on target/linux/ramips/mt7621/base-files/sbin/ubnt_erx_migrate.sh:
```diff
@@ -0,0 +1,86 @@
+
+confirm_migration
+download_image
+
```
Would it work, if you did 
- `prepare_kernel` here,
- On the running system, modified `platform.sh` for edgerouter to something like `"CI_KERNPART="none"; nand_do_upgrade $1"`
- just call `sysupgrade` with the downloaded upgrade package

Just so that you utilise the existing sysupgrade functionality

darkxst replied:
I was working under the assumption, based on ynezz comments above, I can't modify files in `/lib/upgrade`. I do agree it could be simplified by doing so however, by either patching `platform.sh` or `ubnt.sh` on the running system. I guess however the changes will persist even if sysupgrade fails for some reason (which it probably can't since need to use -F)

@ynezz Is something like msandber mentions acceptable to backport to stable release? or do the migration scripts need to be kept standalone as I have done so far?
